api/dao/projectDAO.js

```diff
@@ -0,0 +1,645 @@
+
+        $sort[sortField] = sortDirection;
+
+        queryAggregates.push($sort);
+    }
+
+    // paging
+    queryAggregates.push(
+    {
+        $facet: 
+        {
+            searchResults:
+            [
+                { $skip: pageNumber * pageSize },
+                { $limit: pageSize }
+            ],
+            meta: 
+            [ { $count: 'searchResultsTotal' } ]
+        }
+    });
+
+    // execute the query aggregates
+    let resultSet = await projectModel.aggregate(queryAggregates)
+                                      .collation(collation)
+                                      .exec();
+
+    // sanitize based on roles
+    resultSet = Utils.filterData('Project', resultSet, roles);
+
+    // hateoas
+
+    for(let projectIndex in resultSet[0].searchResults)
+    {
+        let project = resultSet[0].searchResults[projectIndex];
+
+        project = this.projectHateoas(project, roles);
+    }
+
+    return resultSet;
+};
+
+exports.getProject = async function(roles, projectId)
+{
+    let result = mongoose.model('Project').findById(mongoose.Types.ObjectId(projectId));
+    
+    // sanitize based on roles
+    if (result && roles.includes('public'))
+    {
+        delete result.review180Start;
+        delete result.review45Start;
+        delete result.reviewSuspensions;
+        delete result.reviewExtensions;
+    }
+
+    return result;
+};
+
+exports.createProject = async function (user, project)
+{
+    // default project creation is set to 2002 right now for backwards compatibility with other apps that use this api
+    let projectLegislationYear = project.legislationYear ? project.legislationYear : 2002;
+
+    let projectModel = mongoose.model('Project');
+    let newProject;
+    let projectData;
+
+    if (projectLegislationYear == 2018) 
+    {
+      newProject = new projectModel({legislation_2018: project});
+      projectData = newProject.legislation_2018;
+      projectData.legislation = "2018 Environmental Assessment Act";
+    } 
+    else if (projectLegislationYear == 2002) 
+    {
+      newProject = new projectModel({legislation_2002: project});
+      projectData = newProject.legislation_2002;
+      projectData.legislation = "2002 Environmental Assessment Act";
+    } 
+    else if (projectLegislationYear == 1996) 
+    {
+      newProject = new projectModel({legislation_1996: project});
+      projectData = newProject.legislation_1996;
+      projectData.legislation = "1996 Environmental Assessment Act";
+    }
+
+    if (!newProject) 
+    {
+      throw Error('Failed to create new project from project model. Ensure provided Project matches the Project model.');
+    }
+
+    //Need to add this logic to the put because we will only hit a post on a net new project
+    newProject.currentLegislationYear = "legislation_" + projectLegislationYear;
+    newProject.legislationYearList.push(projectLegislationYear);
+
+    projectData.proponent = mongoose.Types.ObjectId(project.proponent);
```

Comment:
I am unsure of the trade off between security and performance, but should we validate that these object IDs exist?